torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2445,43 +2445,50 @@ Tensor det_backward(const Tensor & grad, const Tensor& self, const Tensor& det)
   };
 
   auto nonsingular_case_backward = [&](const Tensor& grad, const Tensor& self, const Tensor& det) -> Tensor {
-    return unsqueeze_multiple(grad * det, {-1, -2}, self.dim()) * self.inverse().transpose(-2, -1);
+    // Derived from Jacobi's formula for partial derivatives, which can be
+    // found at https://en.wikipedia.org/wiki/Jacobi%27s_formula
+    auto det_grad = (unsqueeze_multiple(det, {-1, -2}, self.dim()) * self.inverse().transpose(-2, -1)).conj();
+    return unsqueeze_multiple(grad, {-1, -2}, self.dim()) * det_grad;
   };
 
   if (self.dim() == 2) {
-    if (det.item<double>() == 0) {
+    if (det.eq(0).item<bool>()) {
       return singular_case_backward(grad, self, det);
     } else {
       return nonsingular_case_backward(grad, self, det);
     }
   } else {
-    auto nonzero_det_indices = at::native::toListOfOptionalTensors(at::where(det));
-    c10::optional<Tensor> first_nonzero_det_index = nonzero_det_indices[0];
+    auto nonzero_det_mask = det.ne(0);
 
-    if (first_nonzero_det_index->size(0) == det.numel()) {  // all determinants are nonzero (non-singular)
+    if (nonzero_det_mask.all().item<bool>()) {
       return nonsingular_case_backward(grad, self, det);
     }
 
-    auto zero_det_indices = at::native::toListOfOptionalTensors(at::where(det == 0));
-    c10::optional<Tensor> first_zero_det_index = zero_det_indices[0];
+    auto zero_det_mask = nonzero_det_mask.logical_not();
 
-    if (first_zero_det_index->size(0) == det.numel()) {  // all determinants are zero (singular)
+    if (zero_det_mask.all().item<bool>()) {
       return singular_case_backward(grad, self, det);
     }
 
     Tensor grad_det = at::empty_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
 
-    // invertible case
-    grad_det.index_put_(/*indices=*/nonzero_det_indices,
-                        /*value=*/nonsingular_case_backward(grad.index(nonzero_det_indices),
-                                                            self.index(nonzero_det_indices),
-                                                            det.index(nonzero_det_indices)));
+    auto nonzero_det_list = at::native::toListOfOptionalTensors(nonzero_det_mask);
 
-    // non-invertible case, uses SVD
-    grad_det.index_put_(/*indices=*/zero_det_indices,
-                        /*value=*/singular_case_backward(grad.index(zero_det_indices),
-                                                         self.index(zero_det_indices),
-                                                         det.index(zero_det_indices)));
+    grad_det = grad_det.index_put(
+        nonzero_det_list,
+        nonsingular_case_backward(
+          grad.index(nonzero_det_list),
```

Comment:
In order for double grad to work for complex inputs when we have a mixture of singular and nonsingular matrices, #54562 will have to be merged first. At the moment, `gradgradcheck` fails with `RuntimeError: index does not support automatic differentiation for outputs with complex dtype.`

Reply:
#54562 is merged now!